### PR TITLE
For integer cycling, default icp = 1

### LINKED ISSUE #3610: default to initial cycle point = 1 for integer cycling suites
Tiny thing which kinda bugs me:

```
[scheduling] 
    cycling mode = integer
    [[dependencies]]
        graph = foo
```

This fails because we have not set the initial cycle point.

For `cycling mode = integer` should we set default to `initial cycle point = 1`? *Seems like thats a yes.*

**Pull requests welcome!**

### PR DESCRIPTION
**Pull request by MetRonnie**
These changes close #3610 

**Requirements check-list**
- [X] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [X] Contains logically grouped changes (else tidy your branch by rebase).
- [X] Does not contain off-topic changes (use other PRs for other changes).

- [x] Appropriate tests are included (unit).

- [x] Appropriate change log entry included.

- [X] (master branch) Appropriate doc change made in `suite.py`.

### CODE REVIEW

**Comment by MetRonnie** [comment]
Does this need a test?

**Comment by oliver-sanders** [comment]
Yes, we should have a test for this.

We already have functional tests for integer cycling so the only bit that actually needs testing is the inference of the initial cycle point. That could be done in a unit test.

You would have to take the code that fiddles the ICP out if the `__init__` method into a new method e.g. `adjust_cycle_point(self)` or whatever.

That could then be tested something like this:

```python
# untested
from unitest.mock import Mock

def test_queue_thinggy():
    config = Mock()
    config.cfg = {
        'scheduling': {
            'dependencies': {
                'R1': 'foo'
            }
        }
    }
    cylc.flow.Config.adjust_cycle_point(config)
    assert config.cfg['scheduling']['initial cycle point'] == cylc.flow.cycling.integer.get_point('1')
```

This directory is full of mock-based unit tests - https://github.com/cylc/cylc-flow/tree/master/cylc/flow/tests/main_loop.

**Comment by MetRonnie** [comment]
@oliver-sanders I get a KeyError in `cylc.flow.cycling.loader.get_point_cls()` with my attempted unit test, I don't understand why

<details>
  <summary>traceback</summary>

```
self = <flow.tests.test_config.TestSuiteConfig object at 0x7fe919d615f8>

    def test_integer_cycling_default_initial_point(self):
        """Test that the initial cycle point defaults to 1 for integer cycling
        mode."""
        mocked_config = Mock()
        mocked_config.cfg = OrderedDictWithDefaults({
            'scheduling': OrderedDictWithDefaults({
                'cycling mode': 'integer',
                'initial cycle point': None
                # 'graph': {
                #     'P1': 'foo'
                # }
            })
        })
>       SuiteConfig.process_initial_cycle_point(mocked_config)

test_config.py:231: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../config.py:725: in process_initial_cycle_point
    self.initial_point = get_point(icp).standardise()
../cycling/loader.py:69: in get_point
    return get_point_cls(cycling_type=cycling_type)(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cycling_type = None

    def get_point_cls(cycling_type=None):
        """Return the cylc.flow.cycling.PointBase-derived class we're using."""
        if cycling_type is None:
            cycling_type = DefaultCycler.TYPE
>       return POINTS[cycling_type]
E       KeyError: None

../cycling/loader.py:76: KeyError
```
</details>

**Comment by oliver-sanders** [comment]
Funnily enough I just wrote a pytest fixture about one hour ago which handles that exact problem!

**Comment by MetRonnie** [comment]
Copying & pasting from the PR on this PR, as it might be useful to have this recorded here:

**Oliver:**
> Call the fixture from a test function like so:
> ```python
> def test_mytest(cycling_mode):
>     cycling_mode()
>      # the rest of the test
>```
> FYI, the issue is the way we toggle between different cycling modes.
>
> Rather than passing around a configuration option like a nice same program we instead patch a module variable.
>
> It's nicer (because you don't have to pass around the variable) but a pain in the neck in so many other ways.

**Me:**
> I noticed that before merging this, if I ran `python -m pytest -k test_config.py` (i.e. the whole file), my test would pass, whereas it wouldn't if I just ran my test alone. Is that because the previous tests must have called
https://github.com/cylc/cylc-flow/blob/c360146fe1133bf34e2e5c877f6bae966113c6f3/cylc/flow/cycling/loader.py#L120-L122
>
> So, `DefaultCycler.TYPE`
https://github.com/cylc/cylc-flow/blob/c360146fe1133bf34e2e5c877f6bae966113c6f3/cylc/flow/cycling/loader.py#L57-L61
>
> is sort of like a global variable?

**Oliver:**
> Bingo.
>
> It's a class variable, but exactly the same principle as a global variable, `monkeypatch` will patch it only for the scope of the test function making it safe.
>
> Global variables in modules have odd behaviour, you might see another trick to get around this using lists:
>
> ```python
> MY_LIST = [True]
>
>
> def modify_environment(x):
>    MY_LIST[0] = x
>```

File: cylc/flow/tests/test_config.py [hunk 1]
```diff
@@ -212,6 +216,35 @@ def test_family_inheritance_and_quotes(self):
                 assert 'goodbye_0_major1_minor10' in \
                        config.runtime['descendants']['SOMEFAM']
 
+    def test_missing_initial_cycle_point(self):
```
Comment (oliver-sanders) [inline cylc/flow/tests/test_config.py:219]: These pytests can be run standalone outside of `TestSuiteConfig` (i.e. dedent them one level)

File: cylc/flow/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+@pytest.fixture
+def cycling_mode(monkeypatch):
+    """Set the Cylc cycling mode."""
+    def _cycling_mode(integer=True):
+        monkeypatch.setattr(
+            'cylc.flow.cycling.loader.DefaultCycler.TYPE',
+            (INTEGER_CYCLING_TYPE if integer else ISO8601_CYCLING_TYPE)
+        )
+    return _cycling_mode
```
Comment (kinow) [inline cylc/flow/tests/conftest.py:32]: :rocket:

File: CHANGES.md [hunk 3]
```diff
@@ -33,6 +33,18 @@ namespace, under `cylc.jinja.filters`.
...
+## __cylc-8.0a3 (2020?)__
+
+Fourth alpha release of Cylc 8.
+
+(See note on cylc-8 backward-incompatible changes, above)
+
+### Enhancements
+
+[#3617](https://github.com/cylc/cylc-flow/pull/3617) - For integer cycling mode
+there is now a default initial cycle point of 1.
```
Comment (kinow) [inline CHANGES.md:46]: I thought this could go into alpha 0.2?

Comment (oliver-sanders) [reply CHANGES.md:46]: It can, if approved before the other PRs are merged we can bump it up to a2. We were expecting an earlier release...

**Comment by oliver-sanders** [comment]
Remaining test failure is everyone's favourite unit test so we are good to go.

As a2 has not been released yet and (won't get released until next week) we could slip this into a2 if you like @MetRonnie, just need to edit the CHANGELOG.

**Pull request merged**